src/main/java/com/couchbase/lite/replicator/PullerInternal.java

```diff
@@ -456,10 +459,27 @@ private void queueDownloadedRevision(RevisionInternal rev) {
             }
         }
 
-        if (rev != null && rev.getBody() != null)
-            rev.getBody().compact();
+        final Body body = rev.getBody();
+        if (rev != null && body != null)
+            body.compact();
+
+        synchronized (downloadsToInsert) {
+            globalMemorySize+=body.size();
+        }
 
         downloadsToInsert.queueObject(rev);
+        if (tooManyQueuedObjects()) {
+            Log.d(TAG, "Flushing queue global memory size at " + globalMemorySize);
+            synchronized (downloadsToInsert) {
+                downloadsToInsert.flushAll(true);
```

Comment:
This is a rough hack intended to serve as a clue to how to fix. I was going to go back and figure out if it would be possible to stream data from the network directly to disk lazily as the queue is being processed. I believe, conceptually, this would reduce the memory footprint dramatically.